### PR TITLE
- fix for memory corruption: on windows use _aligned_free() for memor…

### PR DESCRIPTION
**Pull request by guschmue**
…y allocated with aligned_alloc()
- enabled sparse_matmul and immutable_constant_op
- don't print LD_LIBRARY_PATH on error for windows (its null and crashes)
  with this python kernel tests pass (well, some error left related to stl float Inf)

### CODE REVIEW

**Comment by drpngx** [comment]
Jenkins, test this please.

File: tensorflow/contrib/cmake/tf_core_kernels.cmake [hunk 1]
```diff
@@ -85,33 +85,34 @@ list(REMOVE_ITEM tf_core_kernels_srcs ${tf_core_kernels_exclude_srcs})
...
   file(GLOB_RECURSE tf_core_gpu_kernels_exclude_srcs
       # not working on windows yet
       "${tensorflow_source_dir}/tensorflow/core/kernels/avgpooling_op_gpu.cu.cc"
   )
   list(REMOVE_ITEM tf_core_gpu_kernels_srcs ${tf_core_gpu_kernels_exclude_srcs})
-endif(WIN32)
+
+  # if we build for gpu include those  
+  list(APPEND tf_core_kernels_srcs 
+     "${tensorflow_source_dir}/tensorflow/core/kernels/debug_ops.h"
```
Comment (drpngx) [inline tensorflow/contrib/cmake/tf_core_kernels.cmake]: So I guess the windows behavior will be different from the linux one? One has debug ops, only in GPU, and the other one doesn't? We probably want the debug ops to be included (or not) not matter orthogonally to whether we're win32 or gpu.

File: tensorflow/core/common_runtime/gpu/pool_allocator.h [hunk 2]
```diff
@@ -173,7 +173,11 @@ class BasicCPUAllocator : public SubAllocator {
   void* Alloc(size_t alignment, size_t num_bytes) override {
     return port::aligned_malloc(num_bytes, alignment);
   }
+#if defined(PLATFORM_WINDOWS)
```
Comment (drpngx) [inline tensorflow/core/common_runtime/gpu/pool_allocator.h]: there should be a port::aligned_free which is valid for all cases. ASAN or valgrind should be able to flag that.

File: tensorflow/core/kernels/debug_ops.h [hunk 3]
```diff
@@ -127,7 +130,16 @@ class DebugNanCountOp : public OpKernel {
       const T* input_flat = input.template flat<T>().data();
 
       for (int64 i = 0; i < input_shape.num_elements(); ++i) {
-        if (Eigen::numext::isnan(input_flat[i])) {
+#ifdef _MSC_VER
+        // VC complains about ambiguous resolution when isnan is called
+        // with integral values, so we explicitly convert an integral value to
+        // double before calling isnan.
+        typedef std::conditional<std::is_integral<T>::value, double, T>::type R;
+        R value = static_cast<R>(input_flat[i]);
```
Comment (drpngx) [inline tensorflow/core/kernels/debug_ops.h]: ```const R value```

Would just `static_cast<double>(input_flat[i])` work with all types?

File: tensorflow/core/kernels/debug_ops.h [hunk 4]
```diff
@@ -127,7 +130,16 @@ class DebugNanCountOp : public OpKernel {
...
+#else
+        T& value = input_flat[i];
```
Comment (drpngx) [inline tensorflow/core/kernels/debug_ops.h]: const

File: tensorflow/core/kernels/sparse_matmul_op.h [hunk 5]
```diff
@@ -18,6 +18,22 @@ limitations under the License.
 
 #include "third_party/eigen3/Eigen/Core"
 #include "tensorflow/core/platform/types.h"
+#ifdef _MSC_VER
```
Comment (drpngx) [inline tensorflow/core/kernels/sparse_matmul_op.h]: is it possible to find a better home for this, somewhere in platform? Maybe platform/windows/intrinsics_port.h?

**Review by drpngx — CHANGES_REQUESTED** [review]
Nice!

**Comment by drpngx** [comment]
Woops, it doesn't like `std::bind` and other things. Please take a look at the errors:

https://ci.tensorflow.org/job/tensorflow-pull-requests-multijob/2416/

Also, if you're comfortable with git, you could split the commits.

**Comment by guschmue** [comment]
ah, should have tried on linux first. Working on it.

File: tensorflow/contrib/cmake/tf_core_kernels.cmake [hunk 6]
```diff
@@ -85,33 +85,34 @@ list(REMOVE_ITEM tf_core_kernels_srcs ${tf_core_kernels_exclude_srcs})
...
-      "${tensorflow_source_dir}/tensorflow/core/kernels/depthwise_conv_op.cc"  # Cannot find symbol: tensorflow::LaunchConv2DOp<struct Eigen::ThreadPoolDevice, double>::launch(...).
       "${tensorflow_source_dir}/tensorflow/core/kernels/fact_op.cc"
-      "${tensorflow_source_dir}/tensorflow/core/kernels/immutable_constant_op.cc"
-      "${tensorflow_source_dir}/tensorflow/core/kernels/immutable_constant_op.h"
-      "${tensorflow_source_dir}/tensorflow/core/kernels/meta_support.*"
-      "${tensorflow_source_dir}/tensorflow/core/kernels/sparse_matmul_op.cc"
-      "${tensorflow_source_dir}/tensorflow/core/kernels/sparse_matmul_op.h"
       "${tensorflow_source_dir}/tensorflow/core/kernels/*quantiz*.h"
       "${tensorflow_source_dir}/tensorflow/core/kernels/*quantiz*.cc"
+      "${tensorflow_source_dir}/tensorflow/core/kernels/meta_support.*"
```
Comment (guschmue) [inline tensorflow/contrib/cmake/tf_core_kernels.cmake]: there was a mege conflict on meta_support

File: tensorflow/contrib/cmake/tf_core_kernels.cmake [hunk 7]
```diff
@@ -85,33 +85,34 @@ list(REMOVE_ITEM tf_core_kernels_srcs ${tf_core_kernels_exclude_srcs})
...
       "${tensorflow_source_dir}/tensorflow/core/kernels/avgpooling_op.*"
   )
   list(REMOVE_ITEM tf_core_kernels_srcs ${tf_core_kernels_windows_exclude_srcs})
 endif(WIN32)
 
-file(GLOB_RECURSE tf_core_gpu_kernels_srcs
-   "${tensorflow_source_dir}/tensorflow/core/kernels/*.cu.cc"
-   "${tensorflow_source_dir}/tensorflow/contrib/rnn/kernels/*.cu.cc"
-)
+if(WIN32 AND tensorflow_ENABLE_GPU)
```
Comment (guschmue) [inline tensorflow/contrib/cmake/tf_core_kernels.cmake]: it was collecting the files in tf_core_gpu_kernels_srcs but for cpu no target use this variable. That's maybe ok but its risky breaking somebody else using cmake so I figured its better to ifdef it.

File: tensorflow/contrib/cmake/tf_core_kernels.cmake [hunk 8]
```diff
@@ -85,33 +85,34 @@ list(REMOVE_ITEM tf_core_kernels_srcs ${tf_core_kernels_exclude_srcs})
...
 
-if(WIN32)
   file(GLOB_RECURSE tf_core_gpu_kernels_exclude_srcs
       # not working on windows yet
       "${tensorflow_source_dir}/tensorflow/core/kernels/avgpooling_op_gpu.cu.cc"
   )
   list(REMOVE_ITEM tf_core_gpu_kernels_srcs ${tf_core_gpu_kernels_exclude_srcs})
-endif(WIN32)
+
+  # if we build for gpu include those
```
Comment (guschmue) [inline tensorflow/contrib/cmake/tf_core_kernels.cmake]: those don't compile yet, some ambiguous operator error when compiling with msvc. I have a fix but I'm looking for a better fix.

File: tensorflow/core/common_runtime/gpu/pool_allocator.h (hunk 2 above)
Comment (guschmue) [inline tensorflow/core/common_runtime/gpu/pool_allocator.h]: yes, there is  - let me switch to this.

File: tensorflow/core/kernels/debug_ops.h (hunk 3 above)
Comment (guschmue) [inline tensorflow/core/kernels/debug_ops.h]: let me look for  a better way

File: tensorflow/core/kernels/debug_ops.h (hunk 4 above)
Comment (guschmue) [inline tensorflow/core/kernels/debug_ops.h]: ok

File: tensorflow/core/kernels/sparse_matmul_op.h (hunk 5 above)
Comment (guschmue) [inline tensorflow/core/kernels/sparse_matmul_op.h]: fair.

File: tensorflow/core/platform/test.cc [hunk 9]
```diff
@@ -13,6 +13,10 @@ See the License for the specific language governing permissions and
 limitations under the License.
 ==============================================================================*/
 
+#ifdef PLATFORM_WINDOWS
+#include <Windows.h>
```
Comment (guschmue) [inline tensorflow/core/platform/test.cc]: ok

**Comment by guschmue** [comment]
looking over this, you are right: let me split this into a couple of commits. That also helps to get this aligned_free fix in.

**Comment by guschmue** [comment]
closing this one and file some more targeted pr's instead. Going to reference this pr so we don't lose the review feedback.

**Comment by drpngx** [comment]
Please merge with #5207, which will probably create some conflicts on the `meta_support.*`

**Comment by guschmue** [comment]
I closed this pr and split it - too many things in it. I think I'll split it in 4 prs:
1. the memory corruption (addressed your review comments and send a new pr  #5288)
2. enable sparse mat ops (nearly done)
3. enable immutable ops
4. enable debug_op 
I'll make sure sure they are on top of the meta_support.* changes 
Hope that works.

**Comment by drpngx** [comment]
nice, thanks!

**Comment by drpngx** [comment]
BTW, for the `std::bind` thing, we're seeing an error in #5286, possibly the easiest fix would be to use `SparseSlice<TL>*` instead of `auto*`.

**Comment by drpngx** [comment]
Oh, BTW, flipping through some code today, it occurred to me that we should probably `#define WIN32_LEAN_AND_MEAN 1` before including windows.h headers. I usually don't like to have them in header files, only in cpp files, because there a ton of symbols in there.

**Comment by guschmue** [comment]
Ah, lean and mean we have: gets added in CMakeLists.txt to the compiler defines for windows.
Going to try SparseSlice<TL>* shortly. Would love to get rid of the std::bind() without a ifdef.

**Comment by guschmue** [comment]
turns out I did not need the std::bind(). We had made this change before nvidia support vs2015.update3 (we used vs2015.update1 until the official cuda 8.0 came out). Now that the std:bind() made trouble for gcc I started over on sparse_maxmul_op.cc and its fine without std:bind(): builds on linux and windows.

**Comment by drpngx** [comment]
Ah, OK, great to hear! Thanks for making these changes!

**Pull request closed without merging**